datafusion/common/src/utils.rs

```diff
@@ -22,8 +22,16 @@ use arrow::array::ArrayRef;
 use arrow::compute::SortOptions;
 use std::cmp::Ordering;
 
+/// Given column vectors, returns row at `idx`.
+pub fn get_row_at_idx(columns: &[ArrayRef], idx: usize) -> Result<Vec<ScalarValue>> {
+    columns
+        .iter()
+        .map(|arr| ScalarValue::try_from_array(arr, idx))
```

Comment:
What dark magic converts this from an `Vec<Result<ScalarValue>>` to a `Result<Vec<ScalarValue>>`?

Reply:
🙂 `collect` automagically leverages the info in the return type annotation, as if there was a `let` binding or if we had used `collect::<...>`.

Reply:
Longer term, I think there is quite a bit more performance to be gained by avoiding the use of `ScalarValue` here and instead using something more optimized like the RowFormat. This is not something for this PR I am just trying to plan a seed if we need more performance, there is a path